Realm/ObjectServerTests/RLMBSONTests.mm

```diff
@@ -0,0 +1,145 @@
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), bson);
+}
+
+- (void)testIntRoundTrip {
+    auto bson = Bson(int64_t(42));
+    NSNumber *rlm = (NSNumber *)RLMConvertBsonToRLMBSON(bson);
+    XCTAssertEqual(rlm.intValue, 42);
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), bson);
+}
+
+- (void)testBoolRoundTrip {
+    auto bson = Bson(true);
+    NSNumber *rlm = (NSNumber *)RLMConvertBsonToRLMBSON(bson);
+    XCTAssertEqual(rlm.boolValue, true);
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), bson);
+}
+
+- (void)testDoubleRoundTrip {
+    auto bson = Bson(42.42);
+    NSNumber *rlm = (NSNumber *)RLMConvertBsonToRLMBSON(bson);
+    XCTAssertEqual(rlm.doubleValue, 42.42);
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), bson);
+}
+
+- (void)testStringRoundTrip {
+    auto bson = Bson("foo");
+    NSString *rlm = (NSString *)RLMConvertBsonToRLMBSON(bson);
+    XCTAssertEqualObjects(rlm, @"foo");
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), bson);
+}
+
+- (void)testBinaryRoundTrip {
+    auto bson = Bson(std::vector<char>{1, 2, 3});
+    NSData *rlm = (NSData *)RLMConvertBsonToRLMBSON(bson);
+    NSData *d = [[NSData alloc] initWithBytes:(char[]){1, 2, 3} length:3];
+    XCTAssert([rlm isEqualToData: d]);
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), bson);
+}
+
+- (void)testDatetimeMongoTimestampRoundTrip {
+    auto bson = Bson(MongoTimestamp(42, 0));
+    NSDate *rlm = (NSDate *)RLMConvertBsonToRLMBSON(bson);
+    NSDate *d = [[NSDate alloc] initWithTimeIntervalSince1970:42];
+    XCTAssert([rlm isEqualToDate: d]);
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), bson);
+}
+
+- (void)testDatetimeTimestampRoundTrip {
+    auto bson = Bson(realm::Timestamp(42, 0));
+    NSDate *rlm = (NSDate *)RLMConvertBsonToRLMBSON(bson);
+    NSDate *d = [[NSDate alloc] initWithTimeIntervalSince1970:42];
+    XCTAssert([rlm isEqualToDate: d]);
+    // Not an exact round trip since we ignore Timestamp Cocoa side
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), Bson(MongoTimestamp(42, 0)));
+}
+
+- (void)testObjectIdRoundTrip {
+    auto bson = Bson(realm::ObjectId::gen());
+    RLMObjectId *rlm = (RLMObjectId *)RLMConvertBsonToRLMBSON(bson);
+    RLMObjectId *d = [[RLMObjectId alloc] initWithString:rlm.stringValue error:nil];
+    XCTAssert([rlm isEqualTo: d]);
+    XCTAssertEqual(RLMConvertRLMBSONToBson(rlm), bson);
+}
+
+- (void)testDocumentRoundTrip {
+    NSDictionary<NSString *, id<RLMBSON>> *document = @{
+        @"string": @"test string",
+        @"true": @YES,
+        @"false": @NO,
+        @"int": @25,
+        @"int32": @5,
+        @"int64": @10,
+        @"double": @15.0,
+        @"decimal128": [[RLMDecimal128 alloc] initWithString:@"1.2E+10" error:nil],
+        @"minkey": [RLMMinKey new],
+        @"maxkey": [RLMMaxKey new],
+        @"date": [[NSDate alloc] initWithTimeIntervalSince1970: 500],
+        @"nestedarray": @[@[@1, @2], @[@3, @4]],
+        @"nesteddoc": @{@"a": @1, @"b": @2, @"c": @NO, @"d": @[@3, @4]},
```

Comment:
Should this not include `@"e" : @{@"foo": @"bar"}` as another nested element of the document?

Reply:
It 's a nice to have.